source/server/admin/stats_params.h

```diff
@@ -70,6 +70,50 @@ struct StatsParams {
   std::shared_ptr<re2::RE2> re2_filter_;
   Utility::HistogramBucketsMode histogram_buckets_mode_{Utility::HistogramBucketsMode::NoBuckets};
   Http::Utility::QueryParams query_;
+
+  /**
+   * Determines whether a metric should be shown based on the specified query-parameters. This
+   * covers:
+   * ``usedonly``, hidden, and filter.
+   *
+   * @param metric the metric to test
+   * @param name_out if non-null, and the return value is true,
+   *   will contain the metric name. This improves performance because computing the name is
+   *   somewhat expensive, and in some cases it isn't needed.
+   */
+  template <class StatType>
+  bool shouldShowMetric(const StatType& metric, std::string* name_out = nullptr) const {
+    if (used_only_ && !metric.used()) {
+      return false;
+    }
+    if (hidden_ == HiddenFlag::ShowOnly && !metric.hidden()) {
+      return false;
+    }
+    if (hidden_ == HiddenFlag::Exclude && metric.hidden()) {
+      return false;
+    }
+
+    // Performance note: computing the name is expensive.
+    //
+    // In the prometheus case, this is only necessary if doing regex filtering and a regex is
+    // defined (and the function parameter will be nullptr).
+    //
+    // In the non-prometheus case, the name is needed if the metric should be shown, so compute it
+    // once and use it for both regex filtering and in emitting the stat (consumed by caller by
+    // setting `name_out` to non-nullptr).
+    //
+    // For per-endpoint metrics, the name is precomputed and returned as a reference, so storing
+    // it in a std::string causes an extra allocation.
+    if (name_out != nullptr) {
```

Comment:
It's likely not material, but I think sharing this code between prom and non-prom requires 2 extra conditionals per stat. Is that right? Have we compared microbenchmarks?

I think it wouldn't be that hard to reduce that to one extra conditional with the edit right here, at the cost of a small amount of similar looking code in the *then* and *else* sections. WDYT?

Reply:
I agree we don't need a  *new* microbenchmark but there's already a benchmark for the admin stats (test/server/admin:stats_handler_speed_test) . Also it's not just one conditional being changed it's a larger refactor than that.

I suspect it's fine but it's easy enough to check, right?

Reply:
Got it; didn't realize there was already a benchmark. I'll run that and post results.

Reply:
```
new:
# bazel-bin/test/server/admin/stats_handler_speed_test --benchmark_min_time=10
2023-10-16T18:47:25+00:00
Running bazel-bin/test/server/admin/stats_handler_speed_test
Run on (64 X 2294.48 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x64)
  L1 Instruction 32 KiB (x64)
  L2 Unified 4096 KiB (x64)
  L3 Unified 16384 KiB (x1)
Load Average: 0.41, 4.97, 15.98
------------------------------------------------------------------------
Benchmark                              Time             CPU   Iterations
------------------------------------------------------------------------
BM_AllCountersText                   613 ms          613 ms           22
BM_UsedCountersText                 54.7 ms         54.7 ms          261
BM_FilteredCountersText              279 ms          279 ms           50
BM_AllCountersJson                  1009 ms         1009 ms           15
BM_UsedCountersJson                 62.6 ms         62.5 ms          264
BM_FilteredCountersJson              280 ms          280 ms           51
BM_AllCountersPrometheus            5201 ms         5200 ms            3
BM_UsedCountersPrometheus            327 ms          327 ms           42
BM_FilteredCountersPrometheus        759 ms          759 ms           18
BM_HistogramsJson                   25.9 ms         25.9 ms          540

old:
# bazel-bin/test/server/admin/stats_handler_speed_test --benchmark_min_time=10
2023-10-16T19:01:40+00:00
Running bazel-bin/test/server/admin/stats_handler_speed_test
Run on (64 X 2294.48 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x64)
  L1 Instruction 32 KiB (x64)
  L2 Unified 4096 KiB (x64)
  L3 Unified 16384 KiB (x1)
Load Average: 0.76, 15.19, 18.28
------------------------------------------------------------------------
Benchmark                              Time             CPU   Iterations
------------------------------------------------------------------------
BM_AllCountersText                   611 ms          611 ms           23
BM_UsedCountersText                 45.7 ms         45.7 ms          317
BM_FilteredCountersText              250 ms          250 ms           56
BM_AllCountersJson                   907 ms          907 ms           16
BM_UsedCountersJson                 48.4 ms         48.4 ms          299
BM_FilteredCountersJson              249 ms          249 ms           57
BM_AllCountersPrometheus            5211 ms         5210 ms            3
BM_UsedCountersPrometheus            308 ms          308 ms           46
BM_FilteredCountersPrometheus        758 ms          758 ms           19
BM_HistogramsJson                   24.9 ms         24.9 ms          545
```

Reply:
This seems like 10% degradation, no?

Reply:
Those results are quite surprising to me, especially things like `BM_UsedCountersText`.

Reply:
Some tests are 10% slower, some are virtually the same. I don't understand what's going on here. Why would `BM_AllCountersText` not slow down, and several of the others would?

Reply:
AllCountersText doesn't specify a filter.

Reply:
I think this may just be noise in the tests. I re-ran the new version again, with no source changes at all, and now it looks much closer: 

```
# bazel-bin/test/server/admin/stats_handler_speed_test --benchmark_min_time=10
2023-10-16T19:48:31+00:00
Running bazel-bin/test/server/admin/stats_handler_speed_test
Run on (64 X 2294.48 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x64)
  L1 Instruction 32 KiB (x64)
  L2 Unified 4096 KiB (x64)
  L3 Unified 16384 KiB (x1)
Load Average: 0.05, 1.09, 5.86
------------------------------------------------------------------------
Benchmark                              Time             CPU   Iterations
------------------------------------------------------------------------
BM_AllCountersText                   616 ms          616 ms           22
BM_UsedCountersText                 50.3 ms         50.3 ms          290
BM_FilteredCountersText              276 ms          276 ms           48
BM_AllCountersJson                   899 ms          898 ms           15
BM_UsedCountersJson                 51.9 ms         51.9 ms          280
BM_FilteredCountersJson              275 ms          275 ms           52
BM_AllCountersPrometheus            5081 ms         5080 ms            3
BM_UsedCountersPrometheus            310 ms          310 ms           46
BM_FilteredCountersPrometheus        719 ms          719 ms           20
BM_HistogramsJson                   24.9 ms         24.9 ms          550
```

Reply:
It is helpful to repeat it a few times. But can we focus (e.g.) on BM_FilteredCountersText  ? You can run with --benchmark_filter+BM_FilteredCountersText IIRC.

That one is 250ms in the 'old' and 270+ms in the 'new' instances.

Reply:
> AllCountersText doesn't specify a filter.

But it still has to run through all the checks/branches, even if it doesn't execute the re2 part.

Reply:
It could be a a few things. Compiler optimizers are complex.

This new template method is inlined, so my superficial concern of adding extra branches may be solved by the compiler optimizing a constant passed to the inline template function.

OTOH the combined function may be too large to optimize well, and the compiler might have optimized the old one better when it in stats_params.cc.

I do have a question though: is this particular refactoring of shouldShowMetric so it can be shared between prom and non-prom essential to what this PR is doing?

RE the benchmark flakiness, there are some options (use --help to see them) for running multiple iterations and printing std dev, mean, etc.

Reply:
> I do have a question though: is this particular refactoring of shouldShowMetric so it can be shared between prom and non-prom essential to what this PR is doing?

I needed to have the same code [here](https://github.com/envoyproxy/envoy/pull/29709/files#diff-4118abddade5808f6c4be37ae857b876df8748e9a1f55238074deb4e5e9fdba7R203), so I already needed to factor it out into a function. At that point it didn't make sense not to combine them.

Reply:
Ok; worked through this and partially reverted the change. The only part I could figure out that was meaningfully different is that in my version in this PR before today, the `std::string` is first default constructed, then has `move operator=` invoked, then a destructor is run on a temporary. In the original version, the call to `metric.name()` constructs it directly into the string on the stack. I reverted back to this behavior, and now the performance seems to be back where it was (plus/minus margin of error).

```
./bazel-bin/test/server/admin/stats_handler_speed_test --benchmark_min_time=5
2023-10-16T23:38:34+00:00
Running ./bazel-bin/test/server/admin/stats_handler_speed_test
Run on (64 X 2294.48 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x64)
  L1 Instruction 32 KiB (x64)
  L2 Unified 4096 KiB (x64)
  L3 Unified 16384 KiB (x1)
Load Average: 0.84, 0.89, 0.61
------------------------------------------------------------------------
Benchmark                              Time             CPU   Iterations
------------------------------------------------------------------------
BM_AllCountersText                   601 ms          601 ms           12
BM_UsedCountersText                 42.8 ms         42.8 ms          168
BM_FilteredCountersText              259 ms          259 ms           29
BM_AllCountersJson                   881 ms          881 ms            8
BM_UsedCountersJson                 46.4 ms         46.4 ms          150
BM_FilteredCountersJson              251 ms          251 ms           28
BM_AllCountersPrometheus            5043 ms         5043 ms            1
BM_UsedCountersPrometheus            314 ms          314 ms           22
BM_FilteredCountersPrometheus        746 ms          746 ms            9
BM_HistogramsJson                   24.8 ms         24.8 ms          265
```

Reply:
Nice. FWIW if you want to zero in on a particular metric you can use benchmark commands like this:
```
./bazel-bin/test/server/admin/stats_handler_speed_test --benchmark_filter=BM_FilteredCountersText --benchmark_display_aggregates_only=true --benchmark_repetitions=10
2023-10-16T17:41:26-04:00
Running ./bazel-bin/test/server/admin/stats_handler_speed_test
Run on (24 X 4425.78 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x12)
  L1 Instruction 32 KiB (x12)
  L2 Unified 512 KiB (x12)
  L3 Unified 16384 KiB (x4)
Load Average: 2.37, 2.79, 2.89
***WARNING*** CPU scaling is enabled, the benchmark real time measurements may be noisy and will incur extra overhead.
-------------------------------------------------------------------------
Benchmark                               Time             CPU   Iterations
-------------------------------------------------------------------------
BM_FilteredCountersText_mean          232 ms          231 ms           10
BM_FilteredCountersText_median        231 ms          231 ms           10
BM_FilteredCountersText_stddev       1.98 ms         1.97 ms           10
BM_FilteredCountersText_cv           0.86 %          0.85 %            10
[external/com_google_absl/absl/flags/internal/flag.cc : 115] RAW: Restore saved value of envoy_reloadable_features_runtime_initialized to: false
```

Of course the absolute numbers I have are specific to the machine, but it gives you some more statistical confidence by running it a bunch of times.